### ISSUE TITLE
SQL syntax error causes Debug.crash before query ErrorTagger msg is sent back to Elm

### ISSUE THREAD

**Issue by mbylstra** [panosoft/elm-postgres#2]
It seems that if there is an SQL syntax error, the program crashes before we have a chance to handle the ErrorTagger msg. This means we are unable to display or log a useful error message.

Here is some example output from node.js after sending a sql statement with a typo in a column name:

```
Uncaught exception:
 Error: Ran into a `Debug.crash` in module `Postgres` on line 561
The message provided by the code author is:

    Connection Id: 0 is not in state: {
  nextId = 1,
  connections = Dict.fromList [],
  listeners = Dict.fromList [],
  nativeListeners = Dict.fromList []
}
```
I then sprinkled in some debug lines in Postgres.elm and Postgres.js to try to get a better idea of what was going wrong:

```
DEBUG withConnection called for connectionId: 0
DEBUG sql: "\nSELECT to_char(date_monthz, 'YYYY') AS year, to_char(date_month, 'Mon') AS month, num_commits AS total\nFROM (\nSELECT COUNT(*) AS num_commits, date_trunc('month', date) AS date_month\nFROM github_commit \nGROUP BY date_month\nORDER BY date_month ASC\n)  AS unformatted_num_commits_per_month"
DEBUG ErrorQuery: "column \"date_monthz\" does not exist"
DEBUG withConnection called for connectionId: 0
state: { nextId = 1, connections = Dict.fromList [], listeners = Dict.fromList [], nativeListeners = Dict.fromList [] }
Uncaught exception:
 Error: Ran into a `Debug.crash` in module `Postgres` on line 561
The message provided by the code author is:

    Connection Id: 0 is not in state: {
  nextId = 1,
  connections = Dict.fromList [],
  listeners = Dict.fromList [],
  nativeListeners = Dict.fromList []
}
```

As you can see, the "date_monthz" typo causes an SQL syntax error. What seems to be happening is
the `withConnection` task is executed proceeding the SQL Error, but by then the SQL connection has already closed (so the list of connections in State is empty, causing the Debug.crash)
see https://github.com/panosoft/elm-postgres/blob/8b967b03df02eef2a01c4aafbde1664e58b56aec/src/Postgres.elm#L599

Perhaps you could try writing an incorrect SQL statement and see if you have the same issue?

This was my system configuration:
- Ubuntu 14.05 Running as a VM on OSX through Parallels (It's slow)
- PostgresSQL 9.6.1
- Node v6.9.1
- Elm 0.17

Possibly this is a race condition (There may not be an issue if the connection is closed a little later). In a faster set up it might not happen.

**Comment by mbylstra** [comment]
Some further logs showing that the connection has closed before `withConnection` is run:

```
Connect: 0
ConnectionLost: True
ErrorQuery message: `column github_repository.namez does not exist` for SQL: : "\nSELECT github_repository.namez AS repo_name, github_user.login AS user_login, stargazers_count AS num_stars\nFROM github_repository  INNER JOIN github_user  ON github_repository.owner_id = github_user.id\nORDER BY stargazers_count DESC\nLIMIT 20"
Uncaught exception:
 Error: Ran into a `Debug.crash` in module `Postgres` on line 561
The message provided by the code author is:

    Connection Id: 0 is not in state: {
  nextId = 1,
  connections = Dict.fromList [],
  listeners = Dict.fromList [],
  nativeListeners = Dict.fromList []
}

```

**Comment by cscalfani** [comment]
The node `pg` libraries triggers an `error` event in 2 situations. The current version of `elm-postgres` only handles 1.
1. PostGresSQL sends an error message, e.g. when it's stopping. It'll send the Error Object with a message of "terminating connection due to administrator command".
2. The underlying Stream gets an error, e.g. when a SQL syntax error occurs. It'll send an Error String.

_#_1 is currently handled. _#_2 is not and is fixed in 3.0.0 (as of this writing in the next branch).

I've take a conservative approach to the fix which may prove to be a bad decision in the future and I'll have to implement a more explicit check. If an Error Object is provided then the connection is terminated and a ConnectionLost is sent to the App. If an Error String is provided then the error event is ignored.

I'm still working on 3.0.0 which will allow the front-end to use this library via an Authenticating Proxy over Websockets. So the code you'd write on the server to interface to Postgres can also be used on the client with no modifications (sans an additional configuration API call). I'll be releasing a production-ready reference implementation, `PGProxy`, along side of the 3.0.0 release.

I hope to have this released before Thanksgiving.

Since this is not a critical error, I'm inclined to wait until 3.0.0 for the fix. BTW, the API changes are minor. The release is mainly for adding client-side support.
